tools/change_namespace/AddPaths.cpp

```diff
@@ -0,0 +1,119 @@
+#include "ChangeNamespaceImpl.hpp"
+#include "FileHandler.hpp"
+
+#include <iostream>
+#include <regex>
+
+void ChangeNamespaceImpl::add_path(const fs::path& p)
+{
+   fs::path normalized_path = p;
+   normalized_path = normalized_path.lexically_normal();
+   if(fs::exists(m_cppms_path / normalized_path))
+   {
+      if(fs::is_directory(m_cppms_path / normalized_path))
+         add_directory(normalized_path);
+      else
+         add_file(normalized_path);
+   }
+   else
+   {
+      std::cerr << "CAUTION: dependent file " << p.string() << " does not exist." << std::endl;
+      std::cerr << "   Found while scanning file " << m_dependencies[p].string() << std::endl;
+   }
+}
+
+void ChangeNamespaceImpl::add_directory(const fs::path& p)
+{
+   // Parse files and directories
+   fs::directory_iterator i(m_cppms_path / p);
+   fs::directory_iterator j;
+   while(i != j)
+   {
+      std::string s(i->path().string());
+      if(m_cppms_path.string().size())
+         s.erase(0, m_cppms_path.string().size() + 1);
+      fs::path np = s;
+      if(!m_dependencies.count(np)) 
+      {
+         m_dependencies[np] = p; // set up dependency tree	 
+         add_pending_path(np);
+      }
+      ++i;
+   }
+}
+
+void ChangeNamespaceImpl::add_file(const fs::path& p)
+{
+   // If the file is already parsed, return
+   if(m_copy_paths.find(p) != m_copy_paths.end())
+      return;
+
+   // Else add the file to our list
+   m_copy_paths.insert(p);
+
+   // For source files, scan for dependencies:
+   if(is_source_file(p))
+   {
+      add_file_dependencies(p, false);
+   }
+}
+
+void ChangeNamespaceImpl::add_file_dependencies(const fs::path& p, bool scanfile)
+{
+    static const std::regex e(R"(^[[:blank:]]*(?://@bcp[[:blank:]]+([^\n]*)\n)?#[[:blank:]]*include[[:blank:]]*["<]([^">]+)[">])");
+
+    if (!m_dependencies.count(p))
+        m_dependencies[p] = p; // set terminal dependency
+
+    FileHandler handler;
+    if (scanfile)
+        handler.open(p);
+    else
+        handler.open(m_cppms_path / p);
+
+    const int subs[] = { 1, 2 };
+    std::cregex_token_iterator i(handler.begin(), handler.end(), e, subs);
+    std::cregex_token_iterator j;
+
+    while (i != j)
+    {
+        fs::path include_file;
+        try {
+            std::string discard_message = *i;
+            ++i;
+            if (discard_message.size())
```

Comment:
I recommend using an if-with initializer statement here, see https://skebanga.github.io/if-with-initializer/ for details. It helps to tightly scope variable use to within the if statement.